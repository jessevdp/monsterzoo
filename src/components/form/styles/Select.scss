@import '~@local/styles/variables/spacing';
@import '~@local/styles/mixins/form-reset';

.select {
    display: flex;
    flex-direction: column;
    padding: spacing(1) spacing(2) spacing(2);
    position: relative;

    > label {
        font-size: .6rem;
        letter-spacing: .1em;
        text-transform: uppercase;
    }

    > select {
        @include select-reset;
        font-size: 1rem;
        margin-top: spacing(1);
        margin-left: spacing(2);
        text-transform: capitalize;
    }

    > .icon {
        height: .5rem;
        position: absolute;
        top: 50%;
        right: spacing(3);
        transform: translateY(-50%);
        > path {
            fill: var(--ui-contrast-tertiary);
        }
    }

    > select:disabled ~ .icon {
        display: none;
    }
}