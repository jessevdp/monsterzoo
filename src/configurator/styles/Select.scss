@import '~@local/styles/variables/spacing';
@import '~@local/styles/mixins/form-reset';

.select {
    display: flex;
    flex-direction: column;
    padding: spacing(1) spacing(2) spacing(2);

    > label {
        font-size: .6rem;
        letter-spacing: .1em;
        text-transform: uppercase;
    }

    select {
        @include select-reset;
        font-size: 1rem;
        margin-top: spacing(1);
        margin-left: spacing(2);
        text-transform: capitalize;
    }
}