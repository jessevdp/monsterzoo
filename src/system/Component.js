import uuid from 'uuid/v1';
import {
    isFunction,
    isString,
    isObject,
    isArray,
} from '@local/utilities';

export default class Component {
    /**
     * Creates an instance of Component.
     * @memberof Component
     */
    constructor() {
        this.id = uuid();
        this.state = {};
        registerEventListeners(this);
        this._observer = registerDOMMutationObserver(this);
    }

    /**
     * Render the view of the component
     *
     * @returns {string} a string representation of the rendered view
     * @memberof Component
     */
    render() {
        verifyHasView(this);
        const view  = this.view();
        verifyView(view);
        return addComponentId(view, this.id);
    }

    /**
     * Re-render the component to the DOM.
     * If the component has been rendered to the DOM before this method can be used to replace
     * the entire component with an updated version.
     * 
     * @returns {void}
     * @memberof Component
     */
    update() {
        const element = this.getHTMLElement();
        if (element) {
            const newElement = toDOMNode(this.render());
            element.replaceWith(newElement);
        }
    }

    afterDOMUpdate() {
        cleanUpEffects(this);
        runEffects(this);
    }

    /**
     * Mutate the state of the component
     *
     * @param {(object|function)} state the subset of the state to be mutated, or a function that returns this subset
     * @returns {void}
     * @memberof Component
     */
    setState(state) {
        let newState;
        if (isObject(state)) newState = state;
        else if (isFunction(state)) newState = state(this.state);
        this.state = { ...this.state, ...newState };
        this.update();
    }

    /**
     * Register an event listener on the component
     *
     * @param {string} event a string representing the event-type to listen for
     * @param {string} [selector=undefined]
     * @param {function} handler
     * @returns {void}
     * @memberof Component
     */
    on(event, ...args) {
        const handler = args.pop();
        const selector = args.pop();
        const query = `[data-component-id="${this.id}"] ${selector ? selector : ''}`;
        const options = { capture: true, passive: true };

        document.addEventListener(event, (e) => {
            if (containsElement(query, e.target)) handler(e);
        }, options);
    }

    /**
     * Get the HTMLElement for the component by querying the document for an element with the
     * id of the component.
     *
     * @returns {(HTMLElement|null)} Returns the first element with the same component id. Or null if not found.
     * @memberof Component
     */
    getHTMLElement() {
        return document.querySelector(`[data-component-id="${this.id}"]`);
    }

    /**
     * Clean up any side effects that the Component has created
     * 
     * @returns {void}
     * @memberof Component
     */
    cleanup() {
        this._observer.disconnect();
    }
}

/**
 * Check to if a query on the document contains a certain element.
 *
 * @param {string} query 
 * @param {HTMLElement} target
 * @returns {boolean}
 */
function containsElement(query, target) {
    const elements = document.querySelectorAll(query);
    let found = false;
    elements.forEach(element => {
        if (element.contains(target)) found = true;
    });
    return found;
}

/**
 * Create a MutationObserver that calls Component.afterDOMUpdate when the
 * mounted DOM representation of the component, or any of it's sub-nodes
 * are updated.
 * 
 * @returns {void}
 * @param {Component} component
 */
function registerDOMMutationObserver(component) {
    const config = { childList: true, subtree: true };
    const observer = new MutationObserver(mutations => {
        const $component = component.getHTMLElement();
        const wasUpdated = mutations
            .map(mutation => mutation.addedNodes)
            .map(nodeList => [...nodeList])
            .reduce((nodes, subset) => nodes.concat(subset), [])
            .map(node => node.contains($component))
            .reduce((matched, matches) => matched || matches, false);
        if (wasUpdated) component.afterDOMUpdate();
    });
    observer.observe(document, config);
    return observer;
}

function runEffects(component) {
    if (!isFunction(component.effects)) return;

    const cleanup = [];

    function useEffect(effect) {
        const result = effect();
        if (isFunction(result)) cleanup.push(result);
    }

    component.effects(useEffect);
    component._toCleanUp = cleanup;
}

function cleanUpEffects(component) {
    if (isArray(component._toCleanUp)) {
        component._toCleanUp.forEach(cleanup => cleanup());
    }
}

function registerEventListeners(component) {
    if (isFunction(component.events)) component.events();
}

function verifyHasView(component) {
    if (!isFunction(component.view)) {
        throw new Error('Component must implement a [view] function.');
    }
}

function verifyView(view) {
    if (!isString(view)) {
        throw new Error('Component [view] function must return a string.');
    }
    const wrapper = wrapHtmlString(view);
    if (wrapper.children.length !== 1) {
        throw new Error('Component [view] function must return exactly 1 root element');
    }
}

function addComponentId(view, id) {
    const wrapper = wrapHtmlString(view);
    const element = wrapper.children[0];
    element.setAttribute('data-component-id', id);
    return wrapper.innerHTML;
}

function wrapHtmlString(htmlString, element = 'div') {
    const wrapper = document.createElement(element);
    wrapper.innerHTML = htmlString;
    return wrapper;
}

function toDOMNode(view) {
    const wrapper = wrapHtmlString(view);
    return wrapper.children[0];
}
