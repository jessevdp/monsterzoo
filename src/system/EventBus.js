import { isArray } from '@local/utilities/';

export default class EventBus {
    constructor() {
        this.events = {};
    }

    /**
     * Subscribe to a certain event with a handler
     *
     * @param {string} event
     * @param {function} handler
     * @returns {void}
     */
    on(event, handler) {
        if (!isArray(this.events[event])) this.events[event] = [];
        this.events[event].push(handler);
    }

    /**
     * Unsubscribe a handler form a certain event
     *
     * @param {string} event
     * @param {function} handler
     * @returns {void}
     */
    remove(event, handler) {
        if (!isArray(this.events[event])) return;
        const index = this.events[event].indexOf(handler);
        if (index > -1) this.events[event].splice(index, 1);
        if (this.events[event].length <= 0) delete this.events[event];
    }

    /**
     * Emit an event, passing optional data to every listener/handler.
     *
     * @param {string} event
     * @param {*} [args]
     * @returns {void}
     */
    emit(event, ...args) {
        if (!isArray(this.events[event])) return;
        this.events[event].forEach(handler => handler(...args));
    }
}