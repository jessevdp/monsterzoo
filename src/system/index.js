/* eslint-env node */
const system = require('./system');
module.exports = system.default || system;