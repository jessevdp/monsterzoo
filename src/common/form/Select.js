import { Component, renderTemplate } from '@local/system';
import './styles/Select.scss';

export default class Select extends Component {
    constructor(name, label, options = []) {
        super();
        this.setState({
            name,
            label,
            options,
            value: options[0],
        });
    }

    view() {
        const template = ''
            + '<div class="select bg-300">'
                + '<label class="text-secondary">{{label}}</label>'
                + `<select name="{{name}}"${this.isDisabled() ? ' disabled' : ''}>`
                    + '{{#options}}{{{.}}}{{/options}}'
                + '</select>'
                + '<svg class="icon" viewBox="0 0 9 7" xmlns="http://www.w3.org/2000/svg"><path d="M4.5 7L9 0H0z" fill-rule="evenodd"></path></svg>'
            + '</div>';

        const options = this.state.options.map(option => {
            const selected = this.state.value === option ? ' selected' : '';
            return `<option value="${option}"${selected}>${option}</option>`;
        });

        return renderTemplate(template, {...this.state, options});
    }

    events() {
        this.on('change', 'select', () => {
            const $select = this.getHTMLElement('select');
            const isActive = document.activeElement === $select;
            this.setState({ value: $select.value });
            if (isActive) this.getHTMLElement('select').focus();
        });
    }

    isDisabled() {
        return this.state.options.length < 2;
    }


    set value(value) {
        this.setState(state => {
            if (state.options.includes(value)) return { value };
            else throw new RangeError(); 
        });
    }

    get value() {
        return this.state.value;
    }

    set options(options) {
        this.setState(state => {
            const value = options.includes(state.value) ? state.value : options[0];
            return { options, value };
        });
    }

    get options() {
        return this.state.options;
    }
}