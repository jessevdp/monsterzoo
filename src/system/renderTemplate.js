import mustache from 'mustache';
import { isObject, isArray } from '@local/utilities';
import Component from './Component';

/**
 * Render a template string with a given data object.
 *
 * @export
 * @param {string} template the template string
 * @param {object} [data={}] the view-data object
 * @returns {string} the rendered template populated with the data
 */
export default function renderTemplate (template, data = {}) {
    if (typeof template !== 'string') throw new Error('Invalid [template] parameter, expected a string.');
    if (!isObject(data)) throw new Error('Invalid [data] parameter, expected an object.');
    return mustache.render(template, renderComponentsInObject(data));
}

/**
 * Replaces instances of Component with their rendered version
 *
 * @param {object} object
 * @returns {object}
 */
function renderComponentsInObject(object) {
    const newObject = {};
    for (let key in object) {
        const current = object[key];
        if (current instanceof Component) newObject[key] = current.render();
        else if (isObject(current)) newObject[key] = renderComponentsInObject(current);
        else if (isArray(current)) newObject[key] = renderComponentsInArray(current);
        else newObject[key] = object[key];
    }
    return newObject;
}

/**
 * Replaces instances of Component with their rendered version
 *
 * @param {array} array
 * @returns {array}
 */
function renderComponentsInArray(array) {
    return array.map(item => {
        if (item instanceof Component) return item.render();
        else return item;
    });
}
