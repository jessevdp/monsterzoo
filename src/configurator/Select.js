import { Component, renderTemplate } from '@local/system';
import './styles/Select.scss';

export default class Select extends Component {
    constructor(name, label, options) {
        super();
        this.setState({
            name,
            label,
            options,
            value: options[0],
        });
    }

    view() {
        const template = ''
            + '<div class="select bg-300">'
                + '<label class="text-secondary">{{label}}</label>'
                + '<select name="{{name}}">'
                    + '{{#options}}{{{.}}}{{/options}}'
                + '</select>'
            + '</div>';

        const options = this.state.options.map(option => {
            const selected = this.state.value === option ? ' selected' : '';
            return `<option value="${option}"${selected}>${option}</option>`;
        });

        return renderTemplate(template, {...this.state, options});
    }

    events() {
        this.on('change', 'select', () => {
            const $select = this.getHTMLElement('select');
            const isActive = document.activeElement === $select;
            this.setState({ value: $select.value });
            if (isActive) this.getHTMLElement('select').focus();
        });
    }


    set value(value) {
        this.setState({ value });
    }

    get value() {
        return this.state.value;
    }

    set options(options) {
        this.setState({ options });
    }

    get options() {
        return this.state.options;
    }
}