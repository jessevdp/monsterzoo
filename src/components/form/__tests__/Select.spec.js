import Select from '../Select';
import { Component } from '@local/system';

beforeEach(() => {
    document.body.innerHTML = '';
});

it('renders correctly', () => {
    const options = ['foo', 'bar'];
    const select = new Select('name', 'label', options);
    expect(select.view()).toMatchSnapshot();
    select.cleanup();
})

describe('constructor', () => {
    it('sets the [name] on state', () => {
        const name = 'foobar';
        const select = new Select(name, 'label', ['foo']);
        expect(select.state.name).toBe(name);
        select.cleanup();
    })
    it('sets the [label] on state', () => {
        const label = 'label';
        const select = new Select('name', label, ['foo']);
        expect(select.state.label).toBe(label);
        select.cleanup();
    })
    it('sets the [options] on state', () => {
        const options = ['foo', 'bar'];
        const select = new Select('name', 'label', options);
        expect(select.state.options).toBe(options);
        select.cleanup();
    })
    it('defaults the [value] to the first option', () => {
        const options = ['foo', 'bar'];
        const expected = options[0];
        const select = new Select('name', 'label', options);
        expect(select.state.value).toBe(expected);
        select.cleanup();
    })
})

describe('[options] setter', () => {
    it('updates the [options] on state', () => {
        const newOptions = ['foo', 'bar'];
        const select = new Select('name', 'label', ['foo']);
        select.options = newOptions;
        expect(select.state.options).toBe(newOptions);
        select.cleanup();
    })
    it('uses the setState method', () => {
        const newOptions = ['foo', 'bar'];
        const select = new Select('name', 'label', ['foo']);
        const spy = jest.spyOn(Component.prototype, 'setState');
        select.options = newOptions;
        expect(spy).toHaveBeenCalled();
        select.cleanup();
    })
    describe('when the current [value] is not present in the new [options]', () => {
        it('defaults the [value] to the first option', () => {
            // Arrange
            const options = ['baz', 'derp'];
            const newOptions = ['foo', 'bar'];
            const select = new Select('name', 'label', options);
            select.value = options[0];

            // Act
            select.options = newOptions;

            // Assert
            expect(select.state.value).toBe(newOptions[0]);

            // Cleanup
            select.cleanup();
        })
    })
    describe('when the current [value] is present in the new [options]', () => {
        it('keeps the selected value', () => {
            // Arrange
            const options = ['foo', 'bar'];
            const value = options[0];
            const newOptions = ['baz', 'bar', value];
            const select = new Select('name', 'label', options);
            select.value = value;

            // Act
            select.options = newOptions;

            // Assert
            expect(select.state.value).toBe(value);

            // Cleanup
            select.cleanup();
        })
    })
})

describe('[options] getter', () => {
    it('returns the value of state.options', () => {
        const options = ['foo', 'bar'];
        const select = new Select('name', 'label', options);
        expect(select.options).toBe(options);
        select.cleanup();
    })
})

describe('[value] setter', () => {
    it('updates the [value] variable on state', () => {
        const options = ['foo', 'bar'];
        const newValue = options[1];
        const select = new Select('name', 'label', options);
        select.value = newValue;
        expect(select.state.value).toBe(newValue);
        select.cleanup();
    })
    it('updates the selected value of the input box', () => {
        const options = ['foo', 'bar'];
        const newValue = options[1];
        const select = new Select('name', 'label', options);
        document.body.innerHTML = select.render();
        select.value = newValue;
        const $select = document.querySelector('select');
        expect($select.value).toBe(newValue);
        select.cleanup();
    })
    it('uses the setState method', () => {
        const options = ['foo', 'bar'];
        const value = options[1];
        const select = new Select('name', 'label', options);
        const spy = jest.spyOn(Component.prototype, 'setState');
        select.value = value;
        expect(spy).toHaveBeenCalled();
        select.cleanup();
    })
    describe('when the new value is not present in the set of options', () => {
        it('throws a RangeError', () => {
            const options = ['foo', 'bar'];
            const newValue = 'baz';
            const select = new Select('name', 'label', options);
            expect(() => {
                select.value = newValue;
            })
            .toThrowError(RangeError);
            select.cleanup();
        })
    })
})

describe('[value] getter', () => {
    it('returns the value of state.value', () => {
        const options = ['foo', 'bar'];
        const value = options[1];
        const select = new Select('name', 'label', options);
        select.state.value = value;
        expect(select.value).toBe(value);
        select.cleanup();
    })
})

describe('events', () => {
    it('updates [value] state when the value of the select box changes', () => {
        // Arrange
        const options = ['foo', 'bar'];
        const newValue = options[1];
        const select = new Select('name', 'label', options);
        document.body.innerHTML = select.render();
        const $select = document.querySelector('select');

        // Act
        $select.value = newValue;
        $select.dispatchEvent(new Event('change'));

        // Assert
        expect(select.state.value).toBe(newValue);

        // Cleanup
        select.cleanup();
    })
})

describe('isDisabled', () => {
    test.each([
        [[]],
        [['1']],
    ])('returns [true] if there is less than 2 options (%p)', options => {
        const select = new Select('name', 'label', options);
        expect(select.isDisabled()).toBe(true);
        select.cleanup();
    })
    test.each([
        [['1', '2']],
        [['1', '2', '3', '4']],
    ])('returns [false] if there are than 2 or more options (%p)', options => {
        const select = new Select('name', 'label', options);
        expect(select.isDisabled()).toBe(false);
        select.cleanup();
    })
})