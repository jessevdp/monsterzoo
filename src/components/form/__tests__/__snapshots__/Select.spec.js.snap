// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `"<div class=\\"select bg-300\\"><label class=\\"text-secondary\\">label</label><select name=\\"name\\"><option value=\\"foo\\" selected>foo</option><option value=\\"bar\\">bar</option></select><svg class=\\"icon\\" viewBox=\\"0 0 9 7\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M4.5 7L9 0H0z\\" fill-rule=\\"evenodd\\"></path></svg></div>"`;
